fix capa hint focus

muhammad-ammar opened the pull request:
[EDUCATOR-1555](https://openedx.atlassian.net/browse/EDUCATOR-1555)

[SANDBOX](https://studio-video-events-ma.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@36ba6412144b4acf80db1bfcc5da107b?action=new)

muhammad-ammar commented:
@muzaffaryousaf @mushtaqak @Qubad786 This is ready for review.

muhammad-ammar commented:
@cptvitamin This is ready for your review. [SANDBOX](https://studio-video-events-ma.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@36ba6412144b4acf80db1bfcc5da107b?action=new)

cptvitamin commented:
@muhammad-ammar this is pretty good.  The behavior was unexpected on most of the screen reader's I tested.   Would it be possible to move focus to the ```<li>``` element instead of the ```<strong>``` element?  If it's too hard to do this, then your solution is fine.

muhammad-ammar commented:
@cptvitamin No problem. We can move the focus to `<li>`. I will update the code and sandbox tomorrow and ping you again.

I have some questions though. Why screen reader's are showing different behaviours? Is this issue with browsers? Screen readers are not implemented as per a11y standards? OR? 
How a web developer handler all these differences? oh..these are lot of questions :)

cptvitamin commented:
@muhammad-ammar those are some very complicated questions. The short answer is that there is no specification for *how* screen readers should implement their solutions, so there is bound to be some variability in functionality.  In fact, the same was true for HTML until HTML5 which is why users can, and still do, have different experiences on the same website when bouncing between different browsers and different OSs. HTML5 was the first attempt at specifying specifically *how* browsers should *implement* support for features/elements in HTML.  Prior to HTML5, it was left up to the browsers to figure out how they should be implemented.  There is currently work underway for specifying how screen readers should specifically support some of these interactions.  However, it gets even more complicated for accessibility because of the way accessibility is provided on various hardware platforms, operating systems, and browsers.  In fact, there are sometimes multiple methods on a single OS.  To learn more, I recommend you read up on the [Accessibility APIs ](http://whatsock.com/training/) used on various platforms and devices.  

For your specific issue in this ticket, I believe it is being caused because you are sending focus to a child element (<strong>) of a container (<li>).  You will get a  more consistent experience across browsers and screen readers if you send focus to the container element instead of one of its children.

muhammad-ammar commented:
@cptvitamin Thanks for the detailed reply.
I have updated the code as you suggested and also updated the sandbox. We now focus the `<li>`. 
[SANDBOX](https://studio-video-events-ma.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@36ba6412144b4acf80db1bfcc5da107b?action=new)

Qubad786 commented on common/test/acceptance/tests/lms/test_lms_problems.py:
```diff
@@ -144,13 +144,13 @@ def verify_demand_hints(self, first_hint, second_hint, expected_events):
 
         # The hint button rotates through multiple hints
-        problem_page.click_hint()
+        problem_page.click_hint(hint_index=0)
```
`problem_page.click_hint()` will do, as by default `hint_index=0`.

cptvitamin commented:
This works perfectly @muhammad-ammar !  nice work 👍
